### PR TITLE
Call prepare_data once per node in DDP (torchelastic)

### PR DESCRIPTION
**Pull request by ananthsub**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Slurm and elastic training create the training processes per node outside of the lightning context. This means that when the fit function calls prepare_data, the assumption that it's only being called on proc 0 is broken and it gets called for each process.. This fixes PyTorchLightning#1878 by calling prepare_data once per node, depending on the local rank.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
Mind add test for this use case?

**Comment by ananthsub** [comment]
@Borda do you have examples of tests for `prepare_data` that I can look at as a template?

**Comment by Borda** [comment]
Not in a template, but you can mock the class (inherit template) and overview prepare data to do extra call count so in test for multi GPU you can test how many times it was used... Similar as did for some loggers

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -776,10 +776,20 @@ def fit(
...
+        if (
+            not (self.use_ddp or self.use_ddp2)
+            or (self.is_slurm_managing_tasks and int(os.environ["SLURM_LOCALID"]) == 0)
+            # torchelastic or general non_slurm ddp
+            or (
+                "WORLD_SIZE" in os.environ
+                and ("GROUP_RANK" in os.environ or "NODE_RANK" in os.environ)
+                and int(os.environ["LOCAL_RANK"]) == 0
+            )
+        ):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:789]: I have two concerns in terms of code quality:
1. these extra 10 lines of code make the fit method even bigger than it already should be.
2. it is hard to read (and to debug). suggestion: break it down like this:
```python 
condition1 = ...
condition2 = ... 
... 

if condition1 and condition2 and not condition3 ...
```

and choose good names for these variables. how does this sound?

**Comment by williamFalcon** [comment]
See #2166 finishing this PR there

**Pull request closed without merging**